controllers/utils.go

```diff
@@ -132,3 +145,240 @@ func isMergable(extraArgs []string, cmd []string) error {
+	return nil
+}
+
+func multipleRolloutManagersExist(err error) bool {
+	return err.Error() == UnsupportedRolloutManagerConfiguration
+}
+
+func invalidRolloutScope(err error) bool {
+	return err.Error() == UnsupportedRolloutManagerClusterScoped ||
+		err.Error() == UnsupportedRolloutManagerNamespaceScoped
+}
+
+// updateStatusConditionOfRolloutManager calls Set Condition of RolloutManager status
+func updateStatusConditionOfRolloutManager(ctx context.Context, newCondition metav1.Condition, rm *rolloutsmanagerv1alpha1.RolloutManager, k8sClient client.Client, log logr.Logger) error {
+
+	if err := k8sClient.Get(ctx, client.ObjectKeyFromObject(rm), rm); err != nil {
+		log.Error(err, "unable to fetch RolloutManager")
+		return nil
+	}
+
+	changed, newConditions := insertOrUpdateConditionsInSlice(newCondition, rm.Status.Conditions)
+
+	if changed {
+		err := UpdateDeploymentStatusWithFunction(ctx, k8sClient, rm, func(rm *rolloutsmanagerv1alpha1.RolloutManager) {
+			rm.Status.Conditions = newConditions
+		})
+
+		if err != nil {
+			log.Error(err, "unable to update RolloutManager status condition")
+			return err
+		}
+	}
+	return nil
+}
+
+// insertOrUpdateConditionsInSlice is a generic function for inserting/updating metav1.Condition into a slice of []metav1.Condition
+func insertOrUpdateConditionsInSlice(newCondition metav1.Condition, existingConditions []metav1.Condition) (bool, []metav1.Condition) {
+
+	// Check if condition with same type is already set, if Yes then check if content is same,
+	// If content is not same update LastTransitionTime
+
+	index := -1
+	for i, Condition := range existingConditions {
+		if Condition.Type == newCondition.Type {
+			index = i
+			break
+		}
+	}
+
+	now := metav1.Now()
+
+	changed := false
+
+	if index == -1 {
+		newCondition.LastTransitionTime = now
+		existingConditions = append(existingConditions, newCondition)
+		changed = true
+
+	} else if existingConditions[index].Message != newCondition.Message ||
+		existingConditions[index].Reason != newCondition.Reason ||
+		existingConditions[index].Status != newCondition.Status {
+
+		newCondition.LastTransitionTime = now
+		existingConditions[index] = newCondition
+		changed = true
+	}
+
+	return changed, existingConditions
+
+}
+
+func UpdateDeploymentStatusWithFunction(ctx context.Context, k8sClient client.Client, rm *rolloutsmanagerv1alpha1.RolloutManager,
+	mutationFn func(*rolloutsmanagerv1alpha1.RolloutManager)) error {
+
+	return untilSuccess(k8sClient, func(k8sClient client.Client) error {
+
+		// Retrieve the latest version of the RolloutManager resource
+		err := k8sClient.Get(ctx, client.ObjectKeyFromObject(rm), rm)
+		if err != nil {
+			return err
+		}
+
+		// Call the mutation function, to change the RolloutManager
+		mutationFn(rm)
+
+		// Attempt to update the object with the change made by the mutation function
+		err = k8sClient.Status().Update(ctx, rm)
+
+		// Report back the error, if we hit one
+		return err
+	})
```

Comment:
The issue I was talking about is that for failure in updating status condition, there will be some extra time added because of retry mechanism (it will try for given time, i.e 2 min for now, if status is not updated in this duration it will return error anyway), but I think even if we remove this and just reconcile again, then also it will take some time in running all processes again (and make all API calls and create objects again), also logs will be clean else it is filled with update errors. I personally feel retrying status update is better than reconciling all over again, if you disagree I can remove this retry mechanism and just use `.Update().Status()` function.

Reply:
I am using RetryOnConflict instead of PollImmediate.
https://pkg.go.dev/k8s.io/client-go@v0.26.0/util/retry#RetryOnConflict

Reply:
As written above, my preference is to follow the standard operator pattern of returning an error here, rather than using RetryOnConflict/PollImmediate. The issue with using the ROC/PI approach is there is a race condition from when a user makes a change to a resource, to when the status is set:
1) Reconcile is called on resource A, Reconcile starts performing logic
2) Reconcile retrieves a copy of resource A, version 1
2) User makes a change to resource A, producing resource A version 2
3) Reconcile from step 1 finishes (still acting on version 1), and attempts to set status on resource A
    - This fails, as resource A has changed. So, with the ROC/PI approach, Reconciles acquires the new version, and sets the status on that.
4) However, the status that Reconcile is set is only applicable to version 1, since this is the version that Reconcile acted on. 
    - We have thus (temporarily) incorrectly reported a status for an newer version of Resource A, that actually applied to an older version of Resource A.
    
In contrast, with the idiomatic Reconcile-based approach, 3 would fail, and then loop to 1 (where version 2 would be retrieved), solving the problem.

Reply:
Thanks @jparsai! We'll try out this way and see if it causes us any issues, and we can definitely take a second look at our solution once we have some more data from  running and debug the product :crossed_fingers: